proc/updatesearch/updatesearch.py

```diff
@@ -202,33 +213,146 @@ def pipeline_to_xml(self, article, external_metadata=None):
             pipeline_xml.JournalISSNs(),
             pipeline_xml.SubjectAreas(),
             # pipeline_xml.ReceivedCitations(),
+        )
 
-            # Citation IDs
-            pipeline_xml.CitationFK(),
+        xmls = ppl.run([article])
 
-            # Citation Authors
-            pipeline_xml.CitationFKAuthors(),
+        # Constrói elementos <doc> para as referências citadas (<doc>s citação, entities 'citation')
+        # Obtem elementos estrangeiros das referências citadas (para usar no <doc> artigo)
+        cit_xmls, citations_fk = self.get_xmls_citations(article, external_metadata)
 
-            # Citation Normalized Journals' Titles
-            pipeline_xml.CitationFKJournalsExternalData(external_metadata),
+        # Cria elemento raiz
+        add = ET.Element('add')
 
-            # Citation Journals' Titles
-            pipeline_xml.CitationFKJournals(),
+        # Adiciona <doc>s citação na raiz
+        for cit_xml in cit_xmls:
+            add.append(cit_xml)
 
-            # Pipeline Article.Citations
-            pipeline_xml.CitationEntity(external_metadata),
-            pipeline_xml.TearDown()
+        # Adiciona no <doc> artigo as informações estrangeiras das referências citadas
+        for r, xml in xmls:
+            for tag in xml:
+                citations_fk.find('.').append(tag)
+
```

Comment:
@rafaelpezzuto não entendi este trecho. Pode explicar?
`xmls` é resultado de `xmls = ppl.run([article])`, mas qual é o conteúdo de `xmls`? 
Olhando `xmls = ppl.run([article])`, seria o resultado da execução dos pipes, ou seja, seria uma tupla: `(original, arvore_resultante)`. 

Mas ao olhar para o código acima, aparentemente é uma lista de tuplas `(original, arvore_resultante)`, além disso, olhando `for tag in xml`, suponho que `xml` é uma lista?

Reply:
Sim, xmls é na verdade uma lista de tuplas (original, arvore_resultante). Isso ocorre porque na ppl eu não aplico TearDown(). O nome da variável (xmls) não está adequado. 

Veja que eu não fecho a árvore resultante (não uso o TearDown()) e por isso, preciso iterar como demonstrado nessas linhas. Essa iteração é necessária para adicionar na árvore o conteúdo estrangeiro das citações (deve haver maneira mais simples).

Note que aí estamos tratando do <doc> artigo, que contém, além de informações do próprio artigo, também contém valores de todas as citações (apenas autores e periódicos citados). Por isso, não fecho esse doc para poder adicionar nesse mesmo nível os dados das citações.

E no caso das citações, mais a frente, uso a mesma estratégia para (manter a pipe aberta - i.e., não usar TearDown()), em cada citação, para adicionar conteúdo do doc citante.

Posso tentar implementar de uma maneira mais clara. O que acha?

Reply:
Ok, entendido.

Antes de você alterar qualquer coisa, como ficaria este trecho com a implementação "mais clara"?

Reply:
Eu aninharia os dados estrangeiros em docs internos. Então o documento final de um artigo e suas citações seria algo como:

```
<add>
    <doc>
        // dados da citação 1
        ...
        <doc>
            // dados estrangeiros do artigo citante
        </doc>
    </doc>
    <doc>
        // dados da citação 2
        ...
        <doc>
            // dados estrangeiros do artigo citante
        </doc>
    </doc>
    <doc>
        // dados do artigo
        ...
        <doc>
            // dados estrangeiros das citações
        </doc>
    </doc>
</add>
```

Mas ainda não testei se o Solr indexaria como eu espero. E respondendo a sua pergunta, além de outras mudanças, as linhas especificadas ficaram assim:

```
for tag in xml:
    citation_fk.append(tag)
```

Reply:
@rafaelpezzuto acho que a implementação pode ficar assim mesmo, sem alterações. Talvez mudar os nomes das variáveis para ficar mais legível... 
Melhor: como é uma estrutura que se repete, talvez você pudesse transformar em uma função e aí criar teste(s) para ela. Os testes é uma boa abordagem para entender o uso. Que tal?

```python
def add_fields_to_doc(pipeline_results, doc):
    for r, xml in pipeline_results:
        for child in xml:
            doc.find('.').append(child)
```

Reply:
@robertatakenaka ótimo; estou providenciando (i) a mudança dos nomes das variáveis, (ii) a transformação para método e (iii) a criação de teste para esse método.

Grato.